pkg/handler3/handler.go

```diff
@@ -192,7 +243,15 @@ func (o *OpenAPIService) HandleGroupVersion(w http.ResponseWriter, r *http.Reque
 			if err != nil {
 				return
 			}
-			w.Header().Set("Etag", etag)
+			// ETag must be enclosed in double quotes: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag
+			w.Header().Set("Etag", strconv.Quote(etag))
+
+			if hash := r.URL.Query().Get("hash"); hash != "" && hash != etag {
+				u := constructURL(group, etag)
+				http.Redirect(w, r, u, 301)
+				return
+			}
+			applyCacheBusting(w, r, etag)
```

Comment:
Piling up on Alex's comment, a lot of the conditions here are very similar to the conditions in `applyCacheBusting`. Try to remove the `applyCacheBusting` method below and move the code here, I think you'll be able to do a few other simplifications.

Reply:
Since I don't code enough these days, I'll do it for fun. Not addressing the `Vary` comment.
```suggestion
			if hash := r.URL.Query().Get("hash"); hash != "" {
				// The Vary header is required because the Accept header can 
				// change the contents returned. This prevents clients from caching 
				// protobuf as JSON and vice versa.
				// Do you need Vary in the 301?!
				w.Header().Set("Vary", "*")
				if hash != etag {
					u := constructURL(group, etag)
					http.Redirect(w, r, u, 301)
					return
				}
				// Only set these headers when a hash is given.
				w.Header().Set("Cache-Control", "public, immutable")
				// Set the Expires directive to the maximum value of one year from the request,
				// effectively indicating that the cache never expires.
				w.Header().Set("Expires", time.Now().AddDate(1, 0, 0).Format(time.RFC1123))
			}
```

Reply:
So we're not redirecting if no hash is given?